text/0000-variadic-generics.md

```diff
@@ -0,0 +1,444 @@
+types and the possible introduction of padding to propperly align types,
+the type of the value returned from unpacking a `Tupple` should be
+carefully considered.
+
+#### Solution 1
+[solution-1]: #solution-1
+
+When capturing the remaining elements of a `Tupple` with the `...`
+pattern, the result will capture the tail by reference.
+
+```rust
+let some_tuple = (42u8, 65535u16, 0u8);
+let (head, ...tail) = some_tuple; // head: u8, tail: (&u16, &u8)
+
+// If the extracted value of the tuple implements `Copy`
+// then the original `Tuple` has not been moved.
+do_something(some_tuple);
+```
+
+**Limitations**: Capturing the tail by reference going to cause
+us problems later? E.g., what do we do here:
+
+
+```rust
+let some_tuple = (42u8, 0u8, 65535u16, 0u8)
+let (_, ...t0) = some_tuple // t0: (&u8, &u16, &u8)
+let (_, ...t1) = t0 // What is t1? is it (&&u16, &&u8)?
+```
+
+This will increase the number of instructions need to unpack, as a
+new `Tuple` will need to be constructed.
+
+#### Solution 2
+[solution-2]: #solution-2
+
+In theory, instead of merely capturing the tail by reference we could
+construct a new `Tuple` and move each value of the tail to the new
+structure with the added padding.
+
+```rust
+let some_tuple = (42u8, 0u8, 65535u16, 0u8);
+let (_, ...t0) = some_tuple // t0: (u8, u16, u8)
+let (_, ...t1) = t0 // t1: (u16, u8)
+
+do_something(some_tuple); // ERROR: Use of moved value: `some_tuple`
+```
+
+**Limitations**: This will increase the number of instructions need to
+unpack, as a new `Tuple` will need to be constructed.
+
+#### Solution 3
+[solution-3]: #solution-3
+
+The unpacked tuple has some "special sauce" and is not actually allocated.
+A new `TyKind` is created and the compiler essentially transforms the
+following:
+
+```rust
+let some_tuple = (42u8, 0u8, 65535u16, u8);
+let (_, ...tail) = some_tuple;
+
+do_something(...tail);
+```
+
+Into something like the following:
+
+```rust
+let some_tuple = (42u8, 0u8, 65535u16, u8);
+let _ = some_tuple.0
+
+do_something(some_tuple.1, some_tuple.2, some_tuple.3);
+```
+
+Any attempt to use `tail` as a "real" `Tuple` results in a compiler
+error.
+
+**Limitations**: This feels very similar to C\+\+11's Parameter Packs.
+We're creating a new type with its own behavior and missing out on the
+chance to improve the ergonomics of `Tuple`.
+
+#### Solution Discussion
+
+ - Both solutions that allow unpacking a `Tuple` are `O(n)`. Again, we
+   unfortunately can not use a simple `memcpy` due to the possibility
+   of padding changes.
+ - If Solution 1 or 2 is chosen, and the `O(n)` operation of unpacking the
+   original `Tuple` and constructing the tail `Tuple` does not get optimized
+   out by LLVM/MIR when the unpacked tail `Tuple` is not used, a new
+   optimization pass should be written. For example, if the tail `Tuple` is
+   immediately passed to a function, there may be no need to construct
+   the `Tuple` for the `tail`.
```

Comment:
Yes exactly, this is what I would do instead of pushing surprising limitations onto users. In particular, you are trying to lift limitations re. impls and traits but adding whole new ones.